File: polyglot/runtime_template_resolver/py/src/runtime_template_resolver/path_parser.py
from typing import List

def parse_path(path: str) -> List[str]:
    segments = []
    current = []
    in_bracket = False
    in_quote = None

    for char in path:
        if in_quote:
            if char == in_quote:
                in_quote = None
            else:
                current.append(char)
            continue

        if char == '"' or char == "'":
            in_quote = char
            continue

        if char == '[':
            if current:
                segments.append("".join(current))
                current = []
            in_bracket = True
            continue

        if char == ']':
            if current:
                segments.append("".join(current))
                current = []
            in_bracket = False
            continue

        if char == '.' and not in_bracket:
            if current:
                segments.append("".join(current))
                current = []
            continue

        current.append(char)

    if current:
        segments.append("".join(current))
    
    return segments

File: polyglot/runtime_template_resolver/mjs/src/path-parser.ts
/**
 * Parses a path string into segments.
 * Supports dot notation and bracket notation.
 * e.g. "a.b[0]" -> ["a", "b", "0"]
 */
export function parsePath(path: string): string[] {
    const segments: string[] = [];
    let current = '';
    let inBracket = false;
    let inQuote: string | null = null;

    for (let i = 0; i < path.length; i++) {
        const char = path[i];

        if (inQuote) {
            if (char === inQuote) {
                inQuote = null;
            } else {
                current += char;
            }
            continue;
        }

        if (char === '"' || char === "'") {
            inQuote = char;
            continue;
        }

        if (char === '[') {
            if (current) {
                segments.push(current);
                current = '';
            }
            inBracket = true;
            continue;
        }

        if (char === ']') {
            if (current) {
                segments.push(current);
                current = '';
            }
            inBracket = false;
            continue;
        }

        if (char === '.' && !inBracket) {
            if (current) segments.push(current);
            current = '';
            continue;
        }

        current += char;
    }

    if (current) segments.push(current);
    return segments;
}
